### PR TITLE
Concurrency!

### PR DESCRIPTION
**Pull request by sushain97**
Not yet tested

### CODE REVIEW

File: sync.py [hunk 1]
```diff
@@ -199,12 +203,31 @@ def sync_metarepo(clone_dir, name, submodules):
...
+    event_queue.get()
+    logging.info('Starting meta repository update')
+
+    # discard any other piled up events
+    while not event_queue.empty():
+        try:
+            event_queue.get(blocking=False)
+        except queue.Empty:
+            continue
+        event_queue.task_done()
```
Comment (shardulc) [inline sync.py]: Personal taste to move this line inside the `try` clause.

File: sync.py [hunk 2]
```diff
@@ -199,12 +203,31 @@ def sync_metarepo(clone_dir, name, submodules):
...
-def handle_event(args, payload):
+def handle_events(args, event_queue):
+    # wait for an event
+    event_queue.get()
+    logging.info('Starting meta repository update')
+
+    # discard any other piled up events
+    while not event_queue.empty():
+        try:
+            event_queue.get(blocking=False)
```
Comment (shardulc) [inline sync.py]: Or `event_queue.get_nowait()` 😃

File: sync.py [hunk 3]
```diff
@@ -199,12 +203,31 @@ def sync_metarepo(clone_dir, name, submodules):
...
+            continue
+        event_queue.task_done()
+
+    # update the meta repos
     repos = list_repos(args.token)
     repos_by_topic = group_repos_by_topic(repos)
-    for name, topics in REPOS.items():
-        submodules = repos_for_topics(repos_by_topic, topics)
-        sync_metarepo(args.dir, name, submodules)
+    with concurrent.futures.ThreadPoolExecutor() as pool:
```
Comment (shardulc) [inline sync.py]: Isn't it possible that `handle_event` is called again before the previous call's sync pool has finished? (Especially as the default interval is 3 seconds, and git pushes usually take that sort of time.) While that's not really a problem, I think it makes the structure more complicated than if we just synced the submodules in a blocking way.

File: sync.py [hunk 4]
```diff
@@ -199,12 +203,31 @@ def sync_metarepo(clone_dir, name, submodules):
         logging.info('Meta repository %s requires no changes', name)
 
 
-def handle_event(args, payload):
+def handle_events(args, event_queue):
```
Comment (shardulc) [inline sync.py]: I think it makes sense to have `handle_events` as a method in the `RequestHandler` class (and then not have to pass around `event_queue` in three different places).

File: sync.py (hunk 3 above)
Comment (sushain97) [reply sync.py]: No, there's an implicit join at the end of the `with` block:

https://docs.python.org/dev/library/concurrent.futures.html#concurrent.futures.Executor.shutdown

i.e. it does block :)

**Comment by shardulc** [comment]
Looks good to me, except for the optional bit about moving `handle_events` to the handler class. However, this still doesn't *solve* the recursive triggering issue mentioned in #7, but rather makes it so that in practical situations the repos will be updated every 3 seconds all the time. I think the idea of level-2 only triggering level-1 etc. would not be too hard to build into this code.

**Comment by sushain97** [comment]
Ah you're right. I really don't want to introduce any semantic logic here
though.

I think it would be sufficient to introduce an ignore regex to each of the
repos that would match the repositories each repository would not update
for on receipt of an event. The patterns would be very simple and overall
it would be a very simple solution?

**Comment by sushain97** [comment]
Also, I don't think they would sync every 3 seconds. The event that
launches the second sync would result in no commits and therefore no events
of its own if there aren't any real changes.

**Comment by shardulc** [comment]
> no events of its own

I think there would be a push event generated. From the [docs](https://developer.github.com/webhooks/#events):

> `push`: Any Git push to a Repository, including editing tags or branches. Commits via API actions that update references are also counted. This is the default event.

The regex idea works, but for some reason it seems like a hack to me. I've clarified my idea in [the #7 thread](https://github.com/sushain97/apertium-on-github/issues/7#issuecomment-363619589).

**Comment by sushain97** [comment]
> I think there would be a push event generated. From the docs:

Consider the following set of events:
- legitimate commit occurs
- push event
- syncing occurs
- at least one push events
- syncing occurs
- **no push events** because there was no change, `git push` does nothing

There's no infinite loop here, only some CPU cycles wasted.

**Comment by shardulc** [comment]
Hmm... I guess you're right, but we should test this out. If what you say happens then we don't need to worry about this further and can use this code as-is.

**Comment by sushain97** [comment]
I've tested it out :) My server has never infinite looped. I created #7 because there were two sync attempts for a single real push.

File: sync.py [hunk 5]
```diff
@@ -249,24 +229,55 @@ def do_POST(self):
...
+        # discard any other piled up events
+        while not self.event_queue.empty():
+            with contextlib.suppress(queue.Empty):
+                self.event_queue.get_nowait()
+                self.event_queue.task_done()
+
+        # update the meta repos and block until completion
+        repos = list_repos(self.args.token)
+        repos_by_topic = group_repos_by_topic(repos)
+        for metarepos in METAREPOS.topics():
```
Comment (shardulc) [inline sync.py]: `topics()`? 😃

**Review by shardulc — APPROVED** [review]

File: sync.py (hunk 5 above)
Comment (sushain97) [reply sync.py]: o.O

**Pull request merged**